### ISSUE TITLE
Bug: `@typescript-eslint/utils/json-schema` type `JSONSchema4` is wrong

### ISSUE THREAD

**Issue by loynoir** [typescript-eslint/typescript-eslint#7549]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
import { Type } from '@sinclair/typebox'
import type { JSONSchema4 } from '@typescript-eslint/utils/json-schema'
const reproduce: JSONSchema4 = Type.Object({})
void reproduce
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error.

### Actual Result

```txt
Type 'TObject<{}>' is not assignable to type 'JSONSchema4'.
  Type 'TObject<{}>' is not assignable to type 'JSONSchema4ObjectSchema'.
    Types of property 'additionalProperties' are incompatible.
      Type 'TAdditionalProperties' is not assignable to type 'boolean | JSONSchema4 | undefined'
```

### Additional Info

https://github.com/sinclairzx81/typebox/issues/557

### Versions

@typescript-eslint/utils 6.4.1

**Comment by bradzacher** [comment]
I'm not quite sure I understand - how are our types "wrong"? They very, very strictly define JSON Schemas where previous versions were very loose and allowed for all manner of user-land misconfigurations.

We intentionally forked the types so we could simplify them and make them much stricter so we could help people find and fix classes of previously hidden errors.
Using these types we found a number of misconfigured schemas in our own repo.

It's fully expected that our types don't conform to a package using the types from `@types/json-schema`.

For example in the linked issue they list two cases:
```ts
const reproduce: JSONSchema4 = { 
  type: 'object', 
  properties: {}, 
  additionalProperties: {},
};

const reproduce: JSONSchema4 = {};
```

Whilst in the loosest sense both of these objects are valid JSON Schemas - they're also completely non-sensical and a clear programmer error!

In the case of "types that define all valid json schemas" (eg the `@types/json-schema` types) it fully makes sense that you want to be able to describe literally every single valid json schema - even ones that make no semantic sense.

However in our case writing a semantically wrong JSON schema just means that you have a broken rule that users can't configure correctly. It can often lead to holes where you actively lose validation that you assume to be there!!!

TL;DR - nothing broken here - the types intentionally do not match.

**Comment by loynoir** [comment]
@bradzacher

I don't understand why two types is not compact.

But the reproduce example is not valid-normal-real-world-use-case-alike, so, I will give a valid-normal-real-world-use-case-alike reproduce.

```ts
{
  const reproduce: JSONSchema4 = {
    additionalProperties: false,
    type: 'object',
    properties: { a: { type: 'string' }, b: { type: 'string' } },
    required: ['a']
  }

  void reproduce
}
```

But above code maybe **too verbose to read/write**, so, I use `@sinclair/typebox`.

```ts
{
  // @ts-expect-error: TODO
  const reproduce: JSONSchema4 = Type.Strict(
    Type.Object(
      {
        a: Type.String(),
        b: Type.Optional(Type.String())
      },
      { additionalProperties: false }
    )
  )

  void reproduce
}
```

---

Quote

> For example the former declares an object with exactly zero properties - which is useless for a rule config!
> The latter declares an unbounded schema - which accepts anything at all - again useless for a rule config!


@sinclairzx81 
 
What about creating a `StaticSpec`?

```ts
{
  type expected = {
    additionalProperties: false
    type: 'object'
    properties: { a: { type: 'string' }; b: { type: 'string' } }
    required: ['a']
  }

  // @ts-expect-error: TODO: https://github.com/typescript-eslint/typescript-eslint/issues/7549
  const reproduce: expected = Type.StaticSpec(
    Type.Object(
      {
        a: Type.String(),
        b: Type.Optional(Type.String())
      },
      { additionalProperties: false }
    )
  )

  void reproduce
}
```

**Comment by bradzacher** [comment]
`@types/json-schema` defines [`additionalProperties?: boolean | JSONSchema4 | undefined;`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/cc08318d94acc4449b29b66859fb5140ee31c2c8/types/json-schema/index.d.ts#L155)

Where their `JSONSchema4` is [this entire blob of an object](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/cc08318d94acc4449b29b66859fb5140ee31c2c8/types/json-schema/index.d.ts#L62-L241).

We similarly define `additionalProperties`, however our `JSONSchema4` type is a strict union which strictly defines what properties are allowed for each `type`.

So for example - their types allow you to write this code:
```json5
{
  type: "array",
  additionalProperties: true,
}
```

As I mentioned with my example before - this schema is technically completely valid!
A JSON schema library will ignore the `additionalProperties` property because it knows that an array logically cannot have additional properties.

However with our types this schema is disallowed because it makes no sense and likely was a programmer error (i.e. the author likely wanted to use `additionalItems` instead, which will actually work with an `array` schema).

You have this fundamental incompatibility where `@types/json-schema` defines types for all semantically valid JSON schemas, whereas we define types for just the logically correct JSON schemas.

----

I'll put it another way - our types are a strict subtype of `@types/json-schema`.
Here's another way to think about this in more abstract terms
```ts
class Foo { prop = 1; }             // @types/json-schema
class Bar extends Foo { other = 2 } // @typescript-eslint/utils/json-schema

const foo: Foo = new Bar(); // ✅ allowed because Foo is a supertype of Bar
const bar: Bar = new Foo(); // ✅ expected error because Bar is a subtype of Foo
```
[playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG9oAcBO83QLzQEYBuaAX2gsqsoHoboABAFwE80BTCGgKwngDsAtBGAALdgFswAKFCQYAITAZo7AB5N2-ACYwESVPCbiV+AExlodRqw4iMASzRNBnEA-5MaAVyYOQXLwCwmKSMrICEEzQAGaIAFxwiHjQ-OwA7tBKGAAUAJQk1oCg5NBgICDw6eza0ABG7MBg3hDsSUgOMGDQEN4cGLat8DFZyhH8UXXKidkpaZn6+YX0JeocwJo17BhYKvWNza0zHaXd3rUD0ENt0kA)

**Comment by sinclairzx81** [comment]
@bradzacher @loynoir Hi, 

> Whilst in the loosest sense both of these objects are valid JSON Schemas - they're also completely non-sensical and a clear programmer error! For example the former declares an object with exactly zero properties - which is useless for a rule config! The latter declares an unbounded schema - which accepts anything at all - again useless for a rule config!

While I appreciate the view that these schematics may be interpreted as non-sensical, I feel this is largely beside the point. If the goal of JSONSchema4 is to be representative of the specification, it should be supportive of all constructs expressible under that specification (which would extend to supporting the most fundamental schema `{}`)

I don't think it's reasonable to take a strong view of what "is or isn't" considered sensible usage or outright prohibit perceived wrong usage (or imply error on the users part) ... Either JSONSchema4 supports constructs of the Json Schema specification or it doesn't (and if it doesn't, it should probably be documented as such)

### Use Case

The following is a message schema which includes the routing properties `from` and `to` and a permissible `data` property. Such a schema would be typical in a system that forwards messages on behalf of sender and receiver; and where the routing is uninterested in the payload contained within the message.

```typescript
const MessageSchema: JSONSchema4 = { // Type '{}' is not assignable to type 'JSONSchema4' ?
  type: "object",
  required: ['from', 'to', 'data'],
  additionalProperties: false, 
  properties: {
     from: { type: 'string' },
     to: { type: 'string' },
     data: {} // <-- don't care what the data is.
  }
}
```

### TypeScript

The TypeScript analog for the above would be the following; where `unknown` is used to represent the `{}` payload.

```typescript
interface MessageType {
  from: string,
  to: string
  data: unknown
}
```
Just to be clear, I don't know much about the overall goals of this project (I'm merely responding as I've been pinged on this thread), but after reading through the commentary, it does read as though perceptions of what "is and isn't" considered sensible are being used to disallow perfectly reasonable usage of the specification....

Again, just to note that the `{}` schema is [the most fundamental schema expressible under the specification](https://json-schema.org/understanding-json-schema/basics.html#id1) of which all other constraints are applied. As the schema is reconcilable in TypeScript as `unknown`, I do think it makes sense to support it.

All the best
S

**Comment by bradzacher** [comment]
@sinclairzx81 - from the sounds of it we're approaching from two different viewpoints.

In the broad sense of a JSON schema - what you say makes complete sense and I agree with it in many ways.

However in the context of an ESLint rule JSON schema is used to define the allowed config for a rule. This means that statements like "I don't care what this is" is largely non-sensical for an ESLint rule because the rule needs to interpret the config in some way in order to make decisions about how to lint the code.

Accepting any valid JSON value is indeed nonsense for an ESLint rule config. I can fully understand a schema like `{ type: ["string", "number"] }` or similar to allow some flexibility in how the user might pass a value - but definitely not `{}`.

> If the goal of JSONSchema4 is to be representative of the specification, it should be supportive of all constructs expressible under that specification

To be clear - the goal of our types (`@typescript-eslint/utils/json-schema`) is *not* to be representative of **_all_** constructs expressible under v4 of the JSON Schema specification.

Instead the goal of our types is to allow users to express a strict subset of the constructs that will be used to define configuration for ESLint rules.

**Comment by sinclairzx81** [comment]
@bradzacher All good. Understood :)
